docs/spec/connection-handshake/L1_2.md

```diff
@@ -0,0 +1,777 @@
+  if (cAlice.state == INIT || cAlice.state == TRYOPEN) {
+    // Construct the appropriate datagram.
+    d := ConnOpenAck{
+      cBob.remote,                // Alice's party details
+      cBob.version,               // Bob's proposed version
+      constructStateProof(cBob),
+      constructClientProof(cBob)}
+
+    // Push the datagram to Bob's module
+    chains[Alice].push(d)
+  }
+}
+```
+
+###### Step 4.a:
+
+```golang
+// Assume `cAlice` is a connection on chain Alice.
+if (cAlice.state == OPEN) {
+  // Fetch the matching connection end from Alice's side
+  cBob := chains[Bob].getConnection(cAlice.remote.connectionIdentifier)
+  
+  // Handshake ConnOpenAck (step 3) done at Alice.
+  // We will push a `ConnOpenConfirm` to Bob. This is step 4.a.
+  if (cBob.state == TRYOPEN) {
+    // Construct the appropriate datagram.
+    d := ConnOpenConfirm{
+      cBob.local,                 // Bob's party details
+      constructStateProof(cAlice)}
+
+    // Push the datagram to Bob's module
+    chains[Bob].push(d)
+  }
+}
+```
+
+#### Helper functions & application-level predicates
+
+###### `verifyProofs`
+
+```golang
+func verifyProofs(remoteStateProof Proof, 
+                  remoteClientProof Proof,
+                  connectionEnd ConnectionEnd)
+  // Two syntactic check: for each proof, verify that it is authentic.
+  // This return either true (success) or false (failure).
+  // It is also acceptable that this function triggers an abort (e.g., due to
+  // malformed input).
+  t = verifyProofAuthenticity(remoteStateProof, connectionEnd)
+  if t == false {
+    return false
+  }
+  if remoteClientProof != nil {
+    // This return either true (success) or false (failure). Aborting is also
+    // acceptable.
+    t = verifyProofAuthenticity(remoteClientProof, connectionEnd)
+    if t == false {
+      return false
+    }
+  }
+
+  // Semantic checks of both proofs: interpret proof content and verify that
+  // they make sense with respect to connection handshake protocol and our
+  // state.
+  // 1. Verify that the remote party has performed the 'INIT' step for this
+  // specific connectionEnd. This helps guarantee the property
+  // [ICS3-Proto-3-ConnectionIntegrity].
+  // May trigger an abort if the verification fails. Usually returns true
+  // or false.
+  t = verifyRemoteState(remoteStateProof, ConnectionEnd)
+  if t == false {
+    return false
+  }
+  // 2. Verify that the client on the remote party is consistent with my state.
+  // It seems that remoteStateProof is necessary for this step. This also
+  // checks, among others, that remoteClientProof.height <= getCurrentHeight()
+  // cf. [ICS 003].
+  if remoteClientProof != nil {
+    // May trigger an abort if the verification fails. Usually returns true
+    // or false.
+    t = verifyRemoteClient(
+        remoteClientProof [,remoteStateProof], ConnectionEnd)
+    if t == false {
+      return false
+    }
+  }
+  return true
+```
```

Comment:
I think the proof need rework, like we did in TLA spec. Given the current IBC spec, both proofs must be taken at the same height, `proof_height`. For the verification of the consensus proof we need, in addition, the `consensus_height`.